symengine/polys/basic_conversions.cpp

```diff
@@ -0,0 +1,176 @@
+                    one, lcm(*den, *rcp_static_cast<const Rational>(it->second)
+                                        ->get_den()));
+            else
+                gen_set[base] = divnum(one, den);
+        }
+    }
+
+    void bvisit(const Pow &x)
+    {
+        if (is_a<const Integer>(*x.get_exp())) {
+            if (rcp_static_cast<const Integer>(x.get_exp())->is_positive()) {
+                x.get_base()->accept(*this);
+            } else {
+                add_to_gen_set(pow(x.get_base(), minus_one), one);
+            }
+
+        } else if (is_a<const Rational>(*x.get_exp())) {
+
+            RCP<const Basic> base = x.get_base();
+            RCP<const Rational> r
+                = rcp_static_cast<const Rational>(x.get_exp());
+            if (r->is_negative())
+                base = pow(base, minus_one);
+            add_to_gen_set(base, divnum(one, r->get_den()));
+
+        } else {
+            umap_basic_num pow_pairs
+                = _find_gens_poly_pow(x.get_exp(), x.get_base());
+            for (auto it : pow_pairs)
+                add_to_gen_set(pow(x.get_base(), it.first), it.second);
+        }
+    }
+
+    void bvisit(const Add &x)
+    {
+        for (auto it : x.dict_)
+            it.first->accept(*this);
+    }
+
+    void bvisit(const Mul &x)
+    {
+        for (auto it : x.dict_)
+            it.first->accept(*this);
+    }
+
+    void bvisit(const Number &x)
+    {
+        // intentionally blank
+    }
+
+    void bvisit(const Basic &x)
+    {
+        add_to_gen_set(x.rcp_from_this(), one);
+    }
+};
+
+class PolyGeneratorVisitorPow : public BaseVisitor<PolyGeneratorVisitorPow>
+{
+private:
+    // the generators are mul(it.first, it.second) not Pow
+    // the_base is the base of the Pow (whose exp we are currently dealing)
+    // numbers must always be positive and of the form (1/d, d belongs to N)
+    umap_basic_num gen_set;
+    RCP<const Basic> the_base;
+
+public:
+    umap_basic_num apply(const Basic &b, const RCP<const Basic> &base)
+    {
+        the_base = base;
+        b.accept(*this);
+        return gen_set;
+    }
+
+    void bvisit(const Add &x)
+    {
+        if (not x.coef_->is_zero())
+            x.coef_->accept(*this);
+
+        for (auto it : x.dict_) {
+            RCP<const Number> mulx = one, divx = one;
+
+            if (it.second->is_negative())
+                mulx = minus_one;
+
+            if (is_a<const Rational>(*it.second))
+                divx = rcp_static_cast<const Rational>(it.second)->get_den();
+
+            gen_set[mul(mulx, it.first)] = divnum(one, divx);
+        }
+    }
+
+    void bvisit(const Mul &x)
+    {
+        // won't handle cases like 2**((x+1)(x+2))
+        // needs `expand` to have been called
+        RCP<const Number> mulx = one, divx = one;
+
+        if (x.coef_->is_negative())
+            mulx = minus_one;
+
+        if (is_a<const Rational>(*x.coef_))
+            divx = rcp_static_cast<const Rational>(x.coef_)->get_den();
```

Comment:
You can do `divx = divnum(one, static_cast<const Rational &>(x*.coef_).get_den())` and modify Line 143